### PR TITLE
Add quantized training (CPU part)

### PR DESCRIPTION
**Pull request by shiyu1994**
This is to add the quantized training for CPU, which speed up the training. (#5606)

### CODE REVIEW

**Comment by shiyu1994** [comment]
It is weird that the `test_monotone_penalty_max` fails in this commit [6a76fde], which doesn't seem to have any changes that will cause the failure. This should call for further attention to check if the test case fails randomly in some cases. Maybe we should simply make this test case more deterministic by setting the number of threads to be 1 or enable the deterministic flag.

(https://github.com/microsoft/LightGBM/pull/5800/commits/6a76fded10072b0ffcf43d2b0433c8f34a5e85e9)
https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=14615&view=logs&j=c28dceab-947a-5848-c21f-eef3695e5f11&t=fa158246-17e2-53d4-5936-86070edbaacf

**Comment by shiyu1994** [comment]
> @guolinke This is ready for review.

Sorry I think one more test case for distributed quantized training should be added in `test_dask.py`. Please pending the review until that is ready.

File: include/LightGBM/bin.h [hunk 1]
```diff
@@ -56,6 +59,28 @@ inline static void HistogramSumReducer(const char* src, char* dst, int type_size
   }
 }
 
+inline static void Int32HistogramSumReducer(const char* src, char* dst, int type_size, comm_size_t len) {
+  const int64_t* src_ptr = reinterpret_cast<const int64_t*>(src);
+  int64_t* dst_ptr = reinterpret_cast<int64_t*>(dst);
```
Comment (guolinke) [inline include/LightGBM/bin.h:64]: is it okay to use int64_t pointer?

Comment (shiyu1994) [reply include/LightGBM/bin.h:64]: I think the alignments of these buffer vectors is 32 bytes now. So, it should be safe to do this.
https://github.com/microsoft/LightGBM/blob/788e1aa6ae08a14b6bf6f2799b5c4d61426b2079/src/treelearner/parallel_tree_learner.h#L89-L91

**Comment by shiyu1994** [comment]
@guolinke

**Review by guolinke — APPROVED** [review]

**Pull request merged**